### PR TITLE
New default handling of `weights` parameter for `RunLeiden`

### PR DESCRIPTION
**Pull request by andrewwbutler**
This PR addresses https://github.com/satijalab/seurat/issues/3168

When `FindClusters` is called with `algorithm = 4` (leiden algorithm) and `method = 'igraph`, the graph built with `igraph::graph_from_adjacency_matrix()` now always uses `weighted = TRUE`. This changes previous default behavior as the `weights` parameter in `FindClusters` previously determined whether this was set to `TRUE`. The `weights` parameter in `FindClusters` has correspondingly been removed from `FindClusters`. 

To maintain consistency between `method = "igraph"` and `method = "matrix"`, the `mode` parameter in `graph_from_adjacency_matrix` was left as the default (as opposed to setting to `undirected`). For some reason using `mode = "undirected"` here would sometimes give slightly different results. 

@TomKellyGenetics @therealgenna

### CODE REVIEW

**Comment by therealgenna** [comment]
@andrewwbutler That looks good to me - thank you!

**Comment by TomKellyGenetics** [comment]
Looks great @andrewwbutler, only minor point is that scripts which already use `FindClusters` with a `weights` parameter may no longer work.

I recommend deprecating the `weights` argument instead for the next Seurat release.

``` r
#define a function
my_function <- function(x, y = NULL){
  if(is.null(y)) y <- 0
  return(x+y)
}
my_function(2)
#> [1] 2
my_function(2, 3)
#> [1] 5

#remove argument y
my_function <- function(x){
  y <- 0
  return(x+y)
}
my_function(2)
#> [1] 2
my_function(2, 3)
#> Error in my_function(2, 3): unused argument (3)

#deprecate argument y
my_function <- function(x, y = NULL){
  if(!is.null(y)) {
    char <- as.character(deparse(sys.call()))
    .Deprecated(new = gsub(paste(",", y), "", char),
               package="Suerat", 
               old = char)
  }
     y <- 0
  return(x+y)
}
my_function(2)
#> [1] 2
my_function(2, 3)
#> Warning: 'my_function(2, 3)' is deprecated.
#> Use 'my_function(2)' instead.
#> See help("Deprecated") and help("Suerat-deprecated").
#> [1] 2
```

<sup>Created on 2020-07-08 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

File: R/clustering.R [hunk 1]
```diff
@@ -747,7 +736,7 @@ RunLeiden <- function(
     object = input,
     partition_type = partition.type,
     initial_membership = initial.membership,
-    weights = weights,
+    weights = NULL,
```
Comment (TomKellyGenetics) [inline R/clustering.R:739]:
```suggestion
    weights = NULL,
```

Remove in next release (deprecate for now).

File: R/clustering.R [hunk 2]
```diff
@@ -155,7 +152,6 @@ FindClusters.Seurat <- function(
   graph.name = NULL,
   modularity.fxn = 1,
   initial.membership = NULL,
-  weights = NULL,
```
Comment (TomKellyGenetics) [inline R/clustering.R:158]:
```suggestion
  weights = NULL,
```

Keep for next release and remove in following one.

File: R/clustering.R [hunk 3]
```diff
@@ -720,17 +714,12 @@ RunLeiden <- function(
       },
     "igraph" = {
       input <- if (inherits(x = object, what = 'list')) {
-        if (is.null(x = weights)) {
```
Comment (TomKellyGenetics) [inline R/clustering.R:723]:
```suggestion 
     if(!is.null(weights)) {
          char <- as.character(deparse(sys.call()))
          .Deprecated(new = gsub(paste(",", y), "", char),
                     package="Suerat", 
                     old = char)
        }
```

Deprecate `weights` input (give warning that will be removed in future and should not be called in scripts).

**Comment by andrewwbutler** [comment]
Thanks for taking the time to look this over @TomKellyGenetics!  I agree that we should introduce changes that could potentially break peoples scripts in a stepwise (deprecated -> defunct) manner. However, in this case `FindClusters` has `...` as a parameter in the function definition, meaning that even if they included `weights` in script, it should still run without an error. 

We also have an internal function `CheckDots()` that tries to verify and warn when unused parameters are getting swallowed by the `...` parameter (this was a particular concern that we tried to guard against in the transition from v2 to v3). In this scenario, the user would get the following warning:
`Warning: The following arguments are not used: weights`

We could keep the `weights` parameter in the `RunLeiden` function itself but since it is not an exported function, I thought it would be less of a concern to drop it there.

**Comment by TomKellyGenetics** [comment]
PS also addresses: #2574

**Pull request merged**